src/database/database.cpp

```diff
@@ -0,0 +1,1103 @@
+ * cloudprotocol::JournalAlertStorageItf implementation
+ **********************************************************************************************************************/
+
+Error Database::SetJournalCursor(const String& cursor)
+{
+    try {
+        *mSession << "UPDATE config SET cursor = ?;", bind(cursor.CStr()), now;
+    } catch (const Poco::Exception& e) {
+        return AOS_ERROR_WRAP(Error(ErrorEnum::eFailed, e.what()));
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error Database::GetJournalCursor(String& cursor) const
+{
+    try {
+        std::string dbCursor;
+
+        *mSession << "SELECT cursor FROM config;", into(dbCursor), now;
+
+        cursor = dbCursor.c_str();
+    } catch (const Poco::Exception& e) {
+        return AOS_ERROR_WRAP(Error(ErrorEnum::eFailed, e.what()));
+    }
+
+    return ErrorEnum::eNone;
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+RetWithError<bool> Database::TableExist(const std::string& tableName)
+{
+    size_t count {0};
+
+    try {
+        Poco::Data::Statement statement {*mSession};
+
+        statement << "SELECT count(*) FROM sqlite_master WHERE name = ? and type='table'", bind(tableName), into(count);
+
+        if (statement.execute() == 0) {
+            return {false, ErrorEnum::eNotFound};
+        }
+
+    } catch (const Poco::Exception& e) {
+        return {false, AOS_ERROR_WRAP(Error(ErrorEnum::eFailed, e.what()))};
+    }
+
+    return {count > 0, ErrorEnum::eNone};
+}
+
+Error Database::DropAllTables()
+{
+    try {
+        LOG_WRN() << "Dropping all tables";
+
+        *mSession << "DROP TABLE IF EXISTS config;", now;
+        *mSession << "DROP TABLE IF EXISTS network;", now;
+        *mSession << "DROP TABLE IF EXISTS services;", now;
+        *mSession << "DROP TABLE IF EXISTS trafficmonitor;", now;
+        *mSession << "DROP TABLE IF EXISTS layers;", now;
+        *mSession << "DROP TABLE IF EXISTS instances;", now;
+
+    } catch (const Poco::Exception& e) {
+        return AOS_ERROR_WRAP(Error(ErrorEnum::eFailed, e.what()));
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error Database::CreateConfigTable()
+{
+    auto [tableExists, err] = TableExist("config");
+
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    if (tableExists) {
+        uint64_t operationVersion {0};
+
+        Tie(operationVersion, err) = GetOperationVersion();
+        if (!err.IsNone()) {
+            return err;
+        }
+
+        if (operationVersion == sm::launcher::Launcher::cOperationVersion) {
+            return ErrorEnum::eNone;
+        }
+
+        if (err = DropAllTables(); !err.IsNone()) {
+            return AOS_ERROR_WRAP(Error(err, "failed to drop all tables"));
+        }
+    }
+
+    *mSession << "CREATE TABLE config ("
```

Comment:
Shouldn't it be inside try catch as well? Also this function is wrapped by try catch in calling function. Do you need this try catch at all?

Reply:
Moved under try catch
I'd like to keep try catch here, so on exception I'll return Error with more precise details regarding the place error occurred